lib/AST/NameLookup.cpp

```diff
@@ -2780,12 +2794,58 @@ createExtensionGenericParams(ASTContext &ctx,
   return toParams;
 }
 
+CollectedOpaqueReprs swift::collectOpaqueReturnTypeReprs(TypeRepr *r) {
+  class Walker : public ASTWalker {
+    CollectedOpaqueReprs &Reprs;
+
+  public:
+    explicit Walker(CollectedOpaqueReprs &reprs) : Reprs(reprs) {}
+
+    bool walkToTypeReprPre(TypeRepr *repr) override {
+      if (auto opaqueRepr = dyn_cast<OpaqueReturnTypeRepr>(repr))
+        Reprs.push_back(opaqueRepr);
+      return true;
+    }
+  };
+
+  CollectedOpaqueReprs reprs;
+  r->walk(Walker(reprs));
+  return reprs;
+}
+
+CollectedOpaqueReprs swift::collectTypeReprs(TypeRepr *r) {
+  class Walker : public ASTWalker {
+    CollectedOpaqueReprs &Reprs;
+
+  public:
+    explicit Walker(CollectedOpaqueReprs &reprs) : Reprs(reprs) {}
+
+    bool walkToTypeReprPre(TypeRepr *repr) override {
+      if (auto opaqueRepr = dyn_cast<OpaqueReturnTypeRepr>(repr)){
+          Reprs.push_back(opaqueRepr);
+          return false;
+      } else if (auto compositionRepr = dyn_cast<CompositionTypeRepr>(repr)){
+          Reprs.push_back(compositionRepr);
+      } else if (auto existential = dyn_cast<ExistentialTypeRepr>(repr)) {
+          return false;
+      } else if (auto identRepr = dyn_cast<IdentTypeRepr>(repr)) {
+          Reprs.push_back(identRepr);
```

Comment:
Maybe the isProtocol() check can go here?

Reply:
updated!